File: java/arcs/core/storage/keys/DatabaseStorageKey.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.storage.keys

import arcs.core.data.Capabilities
import arcs.core.data.Capability
import arcs.core.storage.StorageKey
import arcs.core.storage.StorageKeyFactory
import arcs.core.storage.StorageKeyProtocol
import arcs.core.storage.StorageKeySpec
import arcs.flags.BuildFlags

/**
 * Default database name for DatabaseDriver usage, and referencing using [DatabaseStorageKey]s.
 */
const val DATABASE_NAME_DEFAULT = "arcs"

/** [StorageKey] implementation for a piece of data managed by the DatabaseDriver. */
sealed class DatabaseStorageKey(protocol: StorageKeyProtocol) : StorageKey(protocol) {
  abstract val unique: String
  abstract val dbName: String

  protected fun checkValidity() {
    require(DATABASE_NAME_PATTERN.matches(dbName)) {
      "$dbName is an invalid database name, must match the pattern: $DATABASE_NAME_PATTERN"
    }
  }

  /** [DatabaseStorageKey] for values to be stored on-disk. */
  data class Persistent(
    override val unique: String,
    override val dbName: String = DATABASE_NAME_DEFAULT
  ) : DatabaseStorageKey(StorageKeyProtocol.Database) {
    /** Short persistent storage key, without an entity schema hash. */
    init {
      checkValidity()
    }

    override fun newKeyWithComponent(component: String): StorageKey {
      return Persistent(component, dbName)
    }

    override fun toKeyString() = "$dbName/$unique"

    override fun toString() = super.toString()

    class Factory : StorageKeyFactory(
      protocol,
      Capabilities(
        listOf(
          Capability.Persistence.ON_DISK,
          Capability.Ttl.ANY,
          Capability.Queryable.ANY,
          Capability.Shareable.ANY
        )
      )
    ) {
      override fun create(options: StorageKeyOptions): StorageKey {
        return Persistent(options.location)
      }
    }

    companion object : StorageKeySpec<Persistent> {
      /** Protocol to be used with the database driver for persistent databases. */
      override val protocol = StorageKeyProtocol.Database

      override fun parse(rawKeyString: String) = parseInternal(rawKeyString, ::Persistent)
    }
  }

  /** [DatabaseStorageKey] for values to be stored in-memory. */
  data class Memory(
    override val unique: String,
    override val dbName: String = DATABASE_NAME_DEFAULT
  ) : DatabaseStorageKey(StorageKeyProtocol.InMemoryDatabase) {

    class Factory : StorageKeyFactory(
      protocol,
      Capabilities(
        listOf(
          Capability.Persistence.IN_MEMORY,
          Capability.Ttl.ANY,
          Capability.Queryable.ANY,
          Capability.Shareable.ANY
        )
      )
    ) {
      override fun create(options: StorageKeyOptions): StorageKey {
        return Memory(options.location)
      }
    }

    init {
      checkValidity()
    }

    override fun newKeyWithComponent(component: String): StorageKey {
      return Memory(component, dbName)
    }

    override fun toKeyString() = "$dbName/$unique"

    override fun toString() = super.toString()

    companion object : StorageKeySpec<Memory> {
      /** Protocol to be used with the database driver for in-memory databases. */
      override val protocol = StorageKeyProtocol.InMemoryDatabase

      override fun parse(rawKeyString: String) = parseInternal(rawKeyString, ::Memory)
    }
  }

  companion object {
    private val DATABASE_NAME_PATTERN = "[a-zA-Z][a-zA-Z0-1_-]*".toRegex()
    private val ENTITY_SCHEMA_HASH_PATTERN = "[a-fA-F0-9]+".toRegex()

    /** Regex for the legacy storage key format, which includes the entity schema hash. */
    private val LEGACY_DB_STORAGE_KEY_PATTERN =
      "^($ENTITY_SCHEMA_HASH_PATTERN)@($DATABASE_NAME_PATTERN)/(.+)\$".toRegex()

    /** Regex for the shortened storage key format, which does not include the schema hash. */
    private val SHORT_DB_STORAGE_KEY_PATTERN =
      "^($DATABASE_NAME_PATTERN)/(.+)\$".toRegex()

    /**
     * Special constant to use when [BuildFlags.STORAGE_KEY_REDUCTION] is enabled and no hashes are
     * required.
     */
    private const val SCHEMA_HASH_NOT_REQUIRED = "SCHEMA_HASH_NOT_REQUIRED"

    /**
     * Parses a [DatabaseStorageKey] from [rawKeyString], invoking [builder] with the arguments
     * `unique` and `dbName`
     */
    private fun <T : DatabaseStorageKey> parseInternal(
      rawKeyString: String,
      builder: (String, String) -> T
    ): T {
      // Try parsing the legacy format first.
      var match = LEGACY_DB_STORAGE_KEY_PATTERN.matchEntire(rawKeyString)
      if (match != null) {
        val dbName = match.groupValues[2]
        val unique = match.groupValues[3]
        return builder(unique, dbName)
      }

      // Try parsing the shortened format.
      match = requireNotNull(SHORT_DB_STORAGE_KEY_PATTERN.matchEntire(rawKeyString)) {
        "Not a valid DatabaseStorageKey: $rawKeyString"
      }
      val dbName = match.groupValues[1]
      val unique = match.groupValues[2]
      return builder(unique, dbName)
    }
  }
}

File: src/runtime/storage/database-storage-key.ts
/**
 * @license
 * Copyright (c) 2020 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {assert} from '../../platform/assert-web.js';
import {StorageKey} from './storage-key.js';
import {Capabilities, Persistence, Encryption, Queryable, Ttl, Shareable} from '../capabilities.js';
import {CapabilitiesResolver} from '../capabilities-resolver.js';
import {StorageKeyFactory, StorageKeyOptions} from '../storage-key-factory.js';

export abstract class DatabaseStorageKey extends StorageKey {
  protected static readonly dbNameDefault = 'arcs';

  constructor(protocol: string,
              public readonly unique: string,
              public readonly entitySchemaHash: string,
              public readonly dbName: string) {
    super(protocol);
    if (!/^[a-zA-Z][a-zA-Z0-1_-]*$/.test(this.dbName)) {
      throw new Error(`Invalid dbName: ${this.dbName}`);
    }
    if (!/^[a-fA-F0-9]+$/.test(this.entitySchemaHash)) {
      throw new Error(`Invalid dbName: ${this.dbName}`);
    }
  }

  toString() {
    return `${this.protocol}://${this.entitySchemaHash}@${this.dbName}/${this.unique}`;
  }

  protected static parseKey(key: string): string[] {
    const match = key.match(/^(db|memdb):\/\/([^@]*)@([^/]+)\/(.*)$/);
    if (!match) {
      throw new Error(`Not a valid DatabaseStorageKey: ${key}.`);
    }
    return match;
  }

  static register() {
    CapabilitiesResolver.registerStorageKeyFactory(new PersistentDatabaseStorageKeyFactory());
    CapabilitiesResolver.registerStorageKeyFactory(new MemoryDatabaseStorageKeyFactory());
  }
}

export class PersistentDatabaseStorageKeyFactory extends StorageKeyFactory {
  get protocol() { return PersistentDatabaseStorageKey.protocol; }

  capabilities(): Capabilities {
    return Capabilities.create([Persistence.onDisk(), Ttl.any(), Queryable.any(), Shareable.any()]);
  }

  create(options: StorageKeyOptions): StorageKey {
    return new PersistentDatabaseStorageKey(options.location(), options.schemaHash);
  }
}

export class MemoryDatabaseStorageKeyFactory extends StorageKeyFactory {
  get protocol() { return MemoryDatabaseStorageKey.protocol; }

  capabilities(): Capabilities {
    return Capabilities.create([Persistence.inMemory(), Ttl.any(), Queryable.any(), Shareable.any()]);
  }

  create(options: StorageKeyOptions): StorageKey {
    return new MemoryDatabaseStorageKey(options.location(), options.schemaHash);
  }
}

export class PersistentDatabaseStorageKey extends DatabaseStorageKey {
  public static readonly protocol = 'db';

  constructor(
      unique: string,
      entitySchemaHash: string,
      dbName: string = DatabaseStorageKey.dbNameDefault) {
    super(PersistentDatabaseStorageKey.protocol, unique, entitySchemaHash, dbName);
  }

  childWithComponent(component: string) {
    return new PersistentDatabaseStorageKey(
        `${this.unique}/${component}`, this.entitySchemaHash, this.dbName);
  }

  static fromString(key: string): PersistentDatabaseStorageKey {
    const [_, protocol, entitySchemaHash, dbName, unique] =
        DatabaseStorageKey.parseKey(key);
    assert(protocol === PersistentDatabaseStorageKey.protocol);
    return new PersistentDatabaseStorageKey(unique, entitySchemaHash, dbName);
  }
}

export class MemoryDatabaseStorageKey extends DatabaseStorageKey {
  public static readonly protocol = 'memdb';

  constructor(
      unique: string,
      entitySchemaHash: string,
      dbName: string = DatabaseStorageKey.dbNameDefault) {
    super(MemoryDatabaseStorageKey.protocol, unique, entitySchemaHash, dbName);
  }

  childWithComponent(component: string) {
    return new MemoryDatabaseStorageKey(
        `${this.unique}/${component}`, this.entitySchemaHash, this.dbName);
  }

  static fromString(key: string): MemoryDatabaseStorageKey {
    const [_, protocol, entitySchemaHash, dbName, unique] =
        DatabaseStorageKey.parseKey(key);
    assert(protocol === MemoryDatabaseStorageKey.protocol);
    return new MemoryDatabaseStorageKey(unique, entitySchemaHash, dbName);
  }
}
